Objects Table Selection improvements

SferaDev opened the pull request:
### :pushpin: References

* **Issue:** Closes #12 

### :memo: Implementation

- Streamline Table Header Checkbox with d2-ui-components SimpleCheckbox instead of mui.

- Get all objects ids on page reload and save them to state.

- When changing pages do not forget previous selection.

- Add a notification messages zone as caption of the objects table.

- Rename selection methods to represent the action they accomplish.

- If selection in all pages is implemented by the list method and all items in current page are selected show a message to select all the ids in all pages. (Note that if already selected the message is not shown).

- If there are selected items in other pages, not visible to the user show a message to clear the selection in all pages.

- In some scenarios both messages can appear and are grouped in the same notification panel.

### 🎨 Screenshots

![image](https://user-images.githubusercontent.com/2181866/56764525-34866b00-67a5-11e9-86ec-44d5780d50f7.png)

![image](https://user-images.githubusercontent.com/2181866/56764540-3d773c80-67a5-11e9-9cf9-7e5151e34aed.png)

![image](https://user-images.githubusercontent.com/2181866/56764549-4405b400-67a5-11e9-95f4-4df2edcad1e1.png)

![image](https://user-images.githubusercontent.com/2181866/56764558-4bc55880-67a5-11e9-83a7-0c56161b449b.png)

![image](https://user-images.githubusercontent.com/2181866/56764572-5253d000-67a5-11e9-8a1a-f1f4fb4f1ff7.png)

### :bookmark_tabs: Others

- Test App: [metadata-synchronization.zip](https://github.com/EyeSeeTea/d2-ui-components/files/3118693/metadata-synchronization.zip)

SferaDev committed:
Objects Table Selection improvements

SferaDev committed:
Fix toggle all selection

SferaDev committed:
Add a new message to clear selection in all pages

SferaDev committed:
Fix selection when filtered data

SferaDev committed:
Update messages

tokland commented on i18n/en.pot:
```diff
@@ -50,6 +50,29 @@ msgstr ""
+msgid "All {{count}} items on this page are selected."
+msgid_plural "All {{count}} items on this page are selected."
+msgstr[0] ""
+msgstr[1] ""
```
So this is how plurals work, good to know! 

To sum it up: When a `i18n.t(msg, {count: N})` is found (`count` is special key), gettext creates a msgid and msgid_plural, and -in the general case, see `Plural-Forms`-, `msgstr[0]` is the singular text and `msgstr[1]` the plural text.

https://www.gnu.org/software/gettext/manual/html_node/Translating-plural-forms.html

tokland commented on i18n/en.pot:
```diff
@@ -50,6 +50,29 @@ msgstr ""
+msgid "There are {{count}} items selected on other pages."
+msgid_plural "There are {{count}} items selected on other pages."
+msgstr[0] ""
+msgstr[1] ""
```
We are not creating custom messages for singular/plural, is that deliberate?

The fr/es translations for those keys seem to be missing, after a `yarn update-po` they appear.

tokland commented on src/objects-table/ObjectsTable.js:
```diff
@@ -99,9 +98,9 @@ class ObjectsTable extends React.Component {
         const { model } = this.props;
         const selectedColumnContents = (
-            <Checkbox
+            <SimpleCheckBox
```
Yep, it makes sense to use the same checkbox everywhere.

tokland commented on src/objects-table/ObjectsTable.js:
```diff
@@ -248,15 +268,63 @@ class ObjectsTable extends React.Component {
+                link: i18n.t("Clear selection"),
+                action: this.clearSelection,
+            });
+        }
```
It's ok as it is, no worries :) But just as a note, to make it functional, you may write:

```
const messages = _.compact([
  condition && message, // or condition ? message : null
  condition2 && message2,
  ...
])
```

Granted, it may not look so nice.

tokland commented on src/objects-table/ObjectsTable.js:
```diff
@@ -130,6 +129,7 @@ class ObjectsTable extends React.Component {
             searchValue: null,
             detailsObject: null,
             selection: new Set(),
```
Selection is cleared if a row is clicked and the detail box is open, is that intentional? Also if the contextual menu for a row is open.

SferaDev replied:
No, that's a bug!! Thanks :)

tokland commented on src/objects-table/ObjectsTable.js:
```diff
@@ -248,15 +268,63 @@ class ObjectsTable extends React.Component {
+
+        if (selectedAllInAllPages) {
+            messages.push({
+                message: i18n.t("There are {{count}} items selected in all pages.", {
```
There is an edge case with a un-intuitive. message  Let's imagine I have all items (2010) selected, then I see: `There are 2010 items selected in all pages`. I unselect one, now I see `There are 1990 items selected on other pages`. That's correct, but maybe we could show also the total? `There are 2009 items selected (1990 in other pages)`. Not sure if this makes sense, just as an idea.

SferaDev replied:
Makes sense

tokland reviewed (commented):
That's a nice feature to have, cool! I didn't found any problem on the selection, it worked as expected. So just some comments:

SferaDev committed:
Make message building more functional

SferaDev committed:
Do not clear selection with active rows change

SferaDev committed:
Improve message detection logic

SferaDev committed:
Add missing condition

SferaDev committed:
Reorder code

SferaDev commented:
- Updated the code to allow plurals where needed (only one phrase actually) and removed the plural pot detection on the rest.

- Message building is now purely functional and has be refactored.

- Conditions are now more robust, previously when filtering we could potentially see an invalid select all pages message.

adrianq committed:
Merge branch 'master' into feature/selection
